docs/dev/design/file-tree-diff.rst

```diff
@@ -0,0 +1,234 @@
+Version file tree diff
+======================
+
+Goals
+-----
+
+- Compare files from two versions to identify the files that have been added, removed, or modified.
+- Provide an API for this feature.
+- Integrate this feature to suggest redirects on files that were removed.
+- Integrate this feature to list the files that changed in a pull request.
+
+Non-goals
+---------
+
+- Replace the `docdiff <https://github.com/readthedocs/addons?tab=readme-ov-file#docdiff>`__ feature from addons.
+  That works on the client side, and it's good for comparing the content of files.
+
+Current problems
+----------------
+
+Currently, when a user opens a PRs, they need to manually search for the files of interest (new and modified files).
+We have a GitHub action that links to the root of the documentation preview, that helps a little, but it's not enough.
+
+When files are removed or renamed, users may not be aware that a redirect may be needed.
+We track 404s in our traffic analytics, but they don't keep track of the version,
+and it may be too late to add a redirect when users are already seeing a 404.
+
+In the past, we haven't implemented those features, because it's hard to map the source files to the generated files,
+since that depends on the build tool and configuration used by the project.
+
+Git providers may already offer a way to compare file trees, but again,
+they work on the source files, and not on the generated files.
+
+All hope was lost for having nice features like this, until now.
+
+Proposed solution
+-----------------
+
+Since redirects and files of interest are related to the generated files.
+Instead of working over the source files, we will work over the generated files, which we have access to.
+
+The key points of this feature are:
+
+- Get the diff of the file tree between two versions.
+- Expose that as an API.
+- Integrate that in PR previews.
+
+Diff between two S3 directories
+-------------------------------
+
+We are already using ``rclone`` to speed up uploads to S3,
+``rclone`` has a command (``rclone check``) to return the diff between two directories.
+For this, it uses the metadata of the files, like size and hash
+(it doesn't download the files).
+
+.. code:: console
+
+   $ ls a
+   changed.txt  new.txt  unchanged.txt
+   $ ls b
+   changed.txt  deleted.txt  unchanged.txt
+   $ rclone check --combined=- /usr/src/app/checkouts/readthedocs.org/a /usr/src/app/checkouts/readthedocs.org/b
+   + new.txt
+   - deleted.txt
+   = unchanged.txt
+   * changed.txt
+
+The result is a list of files with a mark indicating if they were added, removed, or modified.
+The result is easy to parse.
+
+To start, we will only consider HTML files (``--include=*.html``).
+
+Lines changed between two files
+-------------------------------
+
+Once we have the list of files that changed, we can use a tool like ``diff`` to get the lines that changed.
```

Comment:
Yes

Reply:
Having the count of lines changed could be useful to sort the changed files, so it's easy to know which files require the most attention.

Reply:
For example, if the current date or commit is injected into each page, all pages will be marked as changed.

Reply:
> Another idea that could help is to compare the size of the files

This could be a middle ground solution. It's not perfect, but it's a good first approach to sort the file changes 👍🏼 -- We can change the implementation in the future for something more robust if required.

Reply:
Hrm, I feel like we might be poking holes in our own implementation already.

Without some accurate detection of file changes, we'll be giving users a large, noisy list of files that haven't changed (all files in a lot of cases) and this will make this feature hard to use for the user.

Docs could include dynamic changes like this in a number of ways, however some of these might be static length and some variable length.

So it doesn't feel like we can reliably detect this amount of noise in the built docs with just metadata or file sizes. Even with line diffing we can't determine or guarantee what changes are _important_ to the project -- how are we going to decide if content in a line diff can be disregarded? File size detection might miss in both directions as well -- missing legitimate small changes and hitting on variable length dynamic content.

It feels like we need to get quite advanced in our detection of changes to make this reliably usable. This feels a little like some of the work we did with in-doc commenting (which was quantifying _how much_ content chunks changed) and it was really involved. Mapping source to built documentation pages would be far more accurate detection but the mapping part might require built documentation referencing the source file explicitly.

Reply:
Also, another thought, maybe our detection of changes should be some combination of user configuration and _HTML parsing_. Users give us the project's main content selector(s), or we have a reliable default, and we store hashes of this content inside this selector at the end of the build process? However even this might only help use cases where dynamic content comes from the theme templates (commit id, date changed, etc in the footer/header).

Reply:
Yeah, we could use our search indexing code to get a better diff, that already parses all sections on the main content. We already have that information in ES (well, not for PR previews).

Reply:
> We will still give the full list of files that changed, but the sorting will help the user to identify which files are more relevant.

A little bit, but in that UX we're still giving the user a huge list of files that changed, putting the burden on the user to figure out what is important. We should be aiming to give the user a minimal, accurate list of files changed by a PR. If a PR only makes small changes to some files, our UX should point the user to those files. Sorting only seems to be a work around for a number of cases, not a UX we actually want to give to users.

Reply:
I'm thinking something like "quick list of maybe relevant changes" (10 top files deleted/changed/added) and then a "view all changed files" link below.

Reply:
Even with that, if we can't distinguish between a small change an author makes and a small change that the build tool makes, sorting won't accurately give the top 10 files changed. Legitimate small changes wouldn't sort any higher up than many dynamic content changes.

I don't know if parsing HTML and only hashing nested content gets around this issue, but rclone will definitely be subject to this.